### PR TITLE
IOperation params array fixes

### LINKED ISSUE #7342: IArrayCreationExpression gives wrong value for params array
for this code where definition of Format is
Format(string format, params object[] arguments);

string.Format("...", new object[] { 1, 2, 3, 4, 5 });
and
string.Format("...", 1, 2, 3, 4, 5);

params array gives two different values for IArrayCreationExpression.ElementValues.

the former one basically puts ArrayCreationExpression under outer ArrayCreationExpression of element size 1. the latter one, correctly put elements in one ArrayCreationExpression with size 5.

### PR DESCRIPTION
**Pull request by JohnHamby**
IOperation was mangling C# params arguments.

When the params arguments were to be implicitly collected into a params array, the second and succeeding params argument turned into an array creation that was missing the prior matching params arguments. (Correct behavior is for the second and succeeding params arguments to not appear at all, at least for arguments in parameter order.)

When the params arguments were provided as an explicit array, the array argument was being wrapped in an array creation with one element. Correct behavior is for the array argument to be unmodified.

This PR addresses these problems. Fixes #7342

@heejaechang @genlu @nguerrera @srivatsn @mavasani might be interested, if any of them are available.

### CODE REVIEW

**Comment by JohnHamby** [comment]
@dotnet-bot test this please.

**Comment by JohnHamby** [comment]
I apologize for merging unreviewed changes, however small. I'm up against a time barrier.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 1]
```diff
@@ -170,9 +175,15 @@ private static IArgument DeriveArgument(int parameterIndex, int argumentIndex, I
...
-                                : new SimpleArgument(parameters[parameterIndex], argument))
+                            ? (argumentIndex >= parameters.Length - 1 &&
+                               parameters.Length > 0 &&
+                               parameters[parameters.Length - 1].IsParams &&
+                               // An argument that is an array of the appropriate type is not a params argument.
+                               (boundArguments.Length > argumentIndex + 1 ||
+                                argument.Type.TypeKind != TypeKind.Array ||
+                                !argument.Type.Equals(parameters[parameters.Length - 1].Type, true))
+                               ? (IArgument)new Argument(ArgumentKind.ParamArray, parameters[parameters.Length - 1], CreateParamArray(parameters[parameters.Length - 1], boundArguments, argumentIndex))
+                               : new SimpleArgument(parameters[parameterIndex], argument))
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: I feel like this conditional shoud be broken up into `if` blocks.  In the current form it's very hard to follow.

File: src/Compilers/Core/CodeAnalysisTest/Diagnostics/OperationTestAnalyzer.cs [hunk 2]
```diff
@@ -685,4 +685,75 @@ public sealed override void Initialize(AnalysisContext context)
...
+                                     operationContext.ReportDiagnostic(Diagnostic.Create(LongParamsDescriptor, argument.Value.Syntax.GetLocation()));
+                                 }
+                             }
+                         }
+                     }
+                 },
+                 OperationKind.InvocationExpression);
+        }
+
+        static long IntegralValue(object value)
```
Comment (jaredpar) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/OperationTestAnalyzer.cs]: Need `private`

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 1 above)
Comment (genlu) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Since this PR is already merged, I will fix this in #7352 (to avoid any potential merging problem for myself)

File: src/Compilers/Core/CodeAnalysisTest/Diagnostics/OperationTestAnalyzer.cs (hunk 2 above)
Comment (genlu) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/OperationTestAnalyzer.cs]: Will fix in #7352 as mentioned above
